### PR TITLE
[NodeJS] Fix 'this' type for stream options.

### PR DESCRIPTION
**Pull request by rbuckton**
> NOTE: Previous PR: #26837

This changes the node declaration file to set a more correct `this` type in the callbacks for `ReadableOptions`, `WritableOptions`, `DuplexOptions`, and `TransformOptions`. It also aligns the `chunk` type in these stream types to be `any` (to align with the use of `any` in `Readable` and `Writable`).

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] ~Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>~
- [x] Increase the version number in the header if appropriate.
- [x] ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

### CODE REVIEW

File: types/node/index.d.ts [hunk 1]
```diff
@@ -1151,7 +1151,7 @@ declare module "http" {
         constructor();
 
         setTimeout(msecs: number, callback?: () => void): this;
-        destroy(error: Error): void;
+        destroy(error?: Error): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: Is `destroy()`  really needed here? Base class `stream.Writeable` has `destroy()` already.

File: types/node/index.d.ts [hunk 2]
```diff
@@ -2777,7 +2777,7 @@ declare module "net" {
 
         bufferSize: number;
         setEncoding(encoding?: string): this;
-        destroy(err?: any): void;
+        destroy(err?: Error): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: Is `destroy()` really needed here? Base class `stream.Duplex` has `destroy()` already via `stream.Readable`.

File: types/node/index.d.ts [hunk 3]
```diff
@@ -3143,7 +3143,6 @@ declare module "fs" {
 
     export class ReadStream extends stream.Readable {
         close(): void;
-        destroy(): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: Why is `destrory()` removed here? According to docs it should be here https://nodejs.org/dist/latest-v10.x/docs/api/stream.html#stream_readable_destroy_error

File: types/node/index.d.ts [hunk 4]
```diff
@@ -5978,8 +5977,8 @@ declare module "stream" {
             highWaterMark?: number;
             encoding?: string;
             objectMode?: boolean;
-            read?: (this: Readable, size?: number) => any;
-            destroy?: (error: Error | null, callback: (error?: Error) => void) => void;
+            read?(this: Readable, size: number): void;
+            destroy?(this: Readable, error: Error | null, callback: (error?: Error | null) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: I think the `error` in callback here is not optional.

File: types/node/index.d.ts [hunk 5]
```diff
@@ -5997,7 +5996,7 @@ declare module "stream" {
             unshift(chunk: any): void;
             wrap(oldStream: NodeJS.ReadableStream): this;
             push(chunk: any, encoding?: string): boolean;
-            _destroy(error: Error | null, callback: (error?: Error) => void): void;
+            _destroy(error: Error | null, callback: (error?: Error | null) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: I think the error in callback here is not optional.

File: types/node/index.d.ts [hunk 6]
```diff
@@ -6009,83 +6008,83 @@ declare module "stream" {
...
 
         export interface WritableOptions {
             highWaterMark?: number;
             decodeStrings?: boolean;
             objectMode?: boolean;
-            write?: (chunk: any, encoding: string, callback: Function) => any;
-            writev?: (chunks: Array<{ chunk: any, encoding: string }>, callback: Function) => any;
-            destroy?: (error: Error | null, callback: (error?: Error) => void) => void;
-            final?: (callback: (error?: Error) => void) => void;
+            write?(this: Writable, chunk: any, encoding: string, callback: (error?: Error) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: `error` in callback should be `Error | null` see https://nodejs.org/dist/latest-v10.x/docs/api/stream.html#stream_writable_write_chunk_encoding_callback_1

File: types/node/index.d.ts [hunk 7]
```diff
@@ -6009,83 +6008,83 @@ declare module "stream" {
...
+            writev?(this: Writable, chunks: Array<{ chunk: any, encoding: string }>, callback: (error?: Error) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: `Error | null` like in `write`

File: types/node/index.d.ts [hunk 8]
```diff
@@ -6009,83 +6008,83 @@ declare module "stream" {
...
             readonly writableLength: number;
             constructor(opts?: WritableOptions);
-            _write(chunk: any, encoding: string, callback: (err?: Error) => void): void;
-            _writev?(chunks: Array<{ chunk: any, encoding: string }>, callback: (err?: Error) => void): void;
-            _destroy(error: Error | null, callback: (error?: Error) => void): void;
-            _final(callback: Function): void;
-            write(chunk: any, cb?: Function): boolean;
-            write(chunk: any, encoding?: string, cb?: Function): boolean;
+            _write(chunk: any, encoding: string, callback: (error?: Error) => void): void;
+            _writev?(chunks: Array<{ chunk: any, encoding: string }>, callback: (error?: Error) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: again `Error | null` twice

File: types/node/index.d.ts [hunk 9]
```diff
@@ -6009,83 +6008,83 @@ declare module "stream" {
...
+            _destroy(error: Error | null, callback: (error?: Error | null) => void): void;
+            _final(callback: (error?: Error) => void): void;
+            write(chunk: any, cb?: (error?: Error) => void): boolean;
+            write(chunk: any, encoding?: string, cb?: (error?: Error) => void): boolean;
```
Comment (Flarna) [inline types/node/index.d.ts]: once again `Error | null`

File: types/node/index.d.ts [hunk 10]
```diff
@@ -6009,83 +6008,83 @@ declare module "stream" {
...
             setDefaultEncoding(encoding: string): this;
-            end(cb?: Function): void;
-            end(chunk: any, cb?: Function): void;
-            end(chunk: any, encoding?: string, cb?: Function): void;
+            end(cb?: (error?: Error) => void): void;
+            end(chunk: any, cb?: (error?: Error) => void): void;
+            end(chunk: any, encoding?: string, cb?: (error?: Error) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: `Error | null`

File: types/node/index.d.ts [hunk 11]
```diff
@@ -6100,67 +6099,72 @@ declare module "stream" {
...
+            removeListener(event: string | symbol, listener: (...args: any[]) => void): this;
         }
 
         export interface DuplexOptions extends ReadableOptions, WritableOptions {
             allowHalfOpen?: boolean;
             readableObjectMode?: boolean;
             writableObjectMode?: boolean;
+            read?(this: Duplex, size: number): void;
+            write?(this: Duplex, chunk: any, encoding: string, callback: (error?: Error) => void): void;
+            writev?(this: Duplex, chunks: Array<{ chunk: any, encoding: string }>, callback: (error?: Error) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: `Error | null`

File: types/node/index.d.ts [hunk 12]
```diff
@@ -6169,31 +6173,36 @@ declare module "stream" {
...
             readonly writableLength: number;
             constructor(opts?: DuplexOptions);
-            _write(chunk: any, encoding: string, callback: (err?: Error) => void): void;
-            _writev?(chunks: Array<{ chunk: any, encoding: string }>, callback: (err?: Error) => void): void;
-            _destroy(error: Error | null, callback: (error?: Error) => void): void;
-            _final(callback: Function): void;
-            write(chunk: any, cb?: Function): boolean;
-            write(chunk: any, encoding?: string, cb?: Function): boolean;
+            _write(chunk: any, encoding: string, callback: (error?: Error) => void): void;
+            _writev?(chunks: Array<{ chunk: any, encoding: string }>, callback: (error?: Error) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: `Error | null`

File: types/node/index.d.ts [hunk 13]
```diff
@@ -6169,31 +6173,36 @@ declare module "stream" {
...
+            _destroy(error: Error | null, callback: (error?: Error | null) => void): void;
+            _final(callback: (error?: Error) => void): void;
+            write(chunk: any, cb?: (error?: Error) => void): boolean;
```
Comment (Flarna) [inline types/node/index.d.ts]: `Error | null`

File: types/node/index.d.ts [hunk 14]
```diff
@@ -6169,31 +6173,36 @@ declare module "stream" {
...
+            write(chunk: any, encoding?: string, cb?: (error?: Error) => void): boolean;
             setDefaultEncoding(encoding: string): this;
-            end(cb?: Function): void;
-            end(chunk: any, cb?: Function): void;
-            end(chunk: any, encoding?: string, cb?: Function): void;
+            end(cb?: (error?: Error) => void): void;
+            end(chunk: any, cb?: (error?: Error) => void): void;
+            end(chunk: any, encoding?: string, cb?: (error?: Error) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: `Error | null`

File: types/node/index.d.ts [hunk 15]
```diff
@@ -6169,31 +6173,36 @@ declare module "stream" {
...
 
-        type TransformCallback = (err?: Error, data?: any) => void;
+        type TransformCallback = (error?: Error, data?: any) => void;
 
         export interface TransformOptions extends DuplexOptions {
-            transform?: (chunk: any, encoding: string, callback: TransformCallback) => any;
-            flush?: (callback: TransformCallback) => any;
+            read?(this: Transform, size: number): void;
+            write?(this: Transform, chunk: any, encoding: string, callback: (error?: Error) => void): void;
+            writev?(this: Transform, chunks: Array<{ chunk: any, encoding: string }>, callback: (error?: Error) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: `Error | null`

**Review by Flarna — CHANGES_REQUESTED** [review]
see dedicated comments

File: types/node/index.d.ts (hunk 6 above)
Comment (rbuckton) [reply types/node/index.d.ts]: In practice however, `undefined` also works.

Comment (Flarna) [reply types/node/index.d.ts]: Sure, as the usual check in code is something like `if (error)` which covers `null` and `undefined` - but if someone uses `if (error === undefined)` it's wrong.

File: types/node/index.d.ts (hunk 10 above)
Comment (Flarna) [reply types/node/index.d.ts]: Just checked docs and source and you are correct, it's registered as 'finished' event which has no args.

**Comment by Flarna** [comment]
Seems some of my review items caused followup issues. Feel free to revert these `| null` changes as this is not the main intention of this PR.

**Comment by rbuckton** [comment]
@Flarna: those changes are fine, there was just an incompatible module augmentation I missed. I would fix ReadableStream/WritableStream as well but I'd rather just leave that for a different PR.

**Review by Flarna — APPROVED** [review]

**Comment by rbuckton** [comment]
It looks like the remaining issues in the CI build are unrelated. The HTML-related issues were fixed by Microsoft/TypeScript#26161 and will in tonight's "next" build, while the "styled-system" errors look to be due to an error in the "styled-system" definitions and not these changes.

**Comment by rbuckton** [comment]
I've created #27827 as a separate PR to address the issue with "styled-system".

**Comment by rbuckton** [comment]
The remaining build failures are unrelated to this PR. @sheetalkamat, can you review?

**Review by sheetalkamat — APPROVED** [review]

**Comment by Flarna** [comment]
@sheetalkamat @rbuckton Do you know if someone of the DT maintainers fixes these unrelated fails? They effect all node PRs (e.g. #27622).

**Comment by rbuckton** [comment]
@Flarna, we are in the process of fixing this.

**Pull request merged**